File: Week 02/id_011/construct-binary-tree-from-preorder-and-inorder-traversal.go
package algorithm00401

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 || len(inorder) == 0 {
		return nil
	}

	root := &TreeNode{Val: preorder[0]}
	mid := findIndex(inorder, preorder[0])

	root.Left = buildTree(preorder[1:mid+1], inorder[:mid])
	root.Right = buildTree(preorder[mid+1:], inorder[mid+1:])

	return root
}

func findIndex(nums []int, value int) int {
	for i := 0; i < len(nums); i++ {
		if nums[i] == value {
			return i
		}
	}
	return -1
}

File: Week 02/id_081/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
/**
 * ConstructBinaryTreeFromPreorderAndInorderTraversal
 */
public class ConstructBinaryTreeFromPreorderAndInorderTraversal {

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return _buildTree(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }
    
    public TreeNode _buildTree(int[] preorder, int preLeft, int preRight, int[] inorder, int inLeft, int inRight){
        if(preLeft > preRight && inLeft > inRight) return null;
        
        int rootValue = preorder[preLeft];
        int rootIndex = inLeft;
        while(rootValue != inorder[rootIndex]) rootIndex++;
        int leftNodeCount = rootIndex - inLeft;
        
        TreeNode leftNode = _buildTree(preorder, preLeft + 1, preLeft + leftNodeCount, inorder, inLeft, rootIndex - 1);
        TreeNode rightNode = _buildTree(preorder, preLeft + leftNodeCount + 1, preRight, inorder, rootIndex + 1, inRight);
        
        TreeNode root = new TreeNode(rootValue);
        root.left = leftNode;
        root.right = rightNode;
        return root;
    }
}